File: tensorflow/contrib/py2tf/converters/control_flow_test.py
# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Tests for control_flow module."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from tensorflow.contrib.py2tf.converters import control_flow
from tensorflow.contrib.py2tf.converters import converter_test_base
from tensorflow.contrib.py2tf.pyct import compiler
from tensorflow.python.framework import constant_op
from tensorflow.python.ops import control_flow_ops
from tensorflow.python.platform import test


class TestNamer(control_flow.SymbolNamer):

  def new_symbol(self, name_root, used):
    i = 0
    while True:
      name = '%s%d' % (name_root, i)
      if name not in used:
        return name
      i += 1


class ControlFlowTest(converter_test_base.TestCase):

  def test_simple_while(self):

    def test_fn(n):
      i = 0
      s = 0
      while i < n:
        s += i
        i += 1
      return s, i, n

    node = self.parse_and_analyze(test_fn, {})
    node = control_flow.transform(node, TestNamer())
    result = compiler.ast_to_object(node)
    setattr(result, 'tf', control_flow_ops)

    with self.test_session() as sess:
      self.assertEqual((10, 5, 5),
                       sess.run(result.test_fn(constant_op.constant(5))))

  def test_while_single_var(self):

    def test_fn(n):
      while n > 0:
        n -= 1
      return n

    node = self.parse_and_analyze(test_fn, {})
    node = control_flow.transform(node, TestNamer())
    result = compiler.ast_to_object(node)
    setattr(result, 'tf', control_flow_ops)

    with self.test_session() as sess:
      self.assertEqual(0, sess.run(result.test_fn(constant_op.constant(5))))

  def test_simple_if(self):

    def test_fn(n):
      a = 0
      b = 0
      if n > 0:
        a = -n
      else:
        b = 2 * n
      return a, b

    node = self.parse_and_analyze(test_fn, {})
    node = control_flow.transform(node, TestNamer())
    result = compiler.ast_to_object(node)
    setattr(result, 'tf', control_flow_ops)

    with self.test_session() as sess:
      self.assertEqual((-1, 0), sess.run(
          result.test_fn(constant_op.constant(1))))
      self.assertEqual((0, -2),
                       sess.run(result.test_fn(constant_op.constant(-1))))

  def test_if_single_var(self):

    def test_fn(n):
      if n > 0:
        n = -n
      return n

    node = self.parse_and_analyze(test_fn, {})
    node = control_flow.transform(node, TestNamer())
    result = compiler.ast_to_object(node)
    setattr(result, 'tf', control_flow_ops)

    with self.test_session() as sess:
      self.assertEqual(-1, sess.run(result.test_fn(constant_op.constant(1))))


if __name__ == '__main__':
  test.main()

File: tensorflow/core/graph/control_flow_test.cc
/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/graph/control_flow.h"

#include <string>
#include <vector>

#include "tensorflow/cc/ops/standard_ops.h"
#include "tensorflow/cc/ops/while_loop.h"
#include "tensorflow/core/lib/core/status_test_util.h"
#include "tensorflow/core/lib/strings/str_util.h"
#include "tensorflow/core/platform/test.h"

namespace tensorflow {
namespace {
Status LessThanTenCond(const Scope& scope, const std::vector<Output>& inputs,
                       Output* output) {
  *output = ops::Less(scope, inputs[0], 10);
  return scope.status();
}

Status AddOneBody(const Scope& scope, const std::vector<Output>& inputs,
                  std::vector<Output>* outputs) {
  outputs->push_back(ops::AddN(scope, {inputs[0], 1}));
  return scope.status();
}

Status NestedLoopBody(const Scope& scope, const std::vector<Output>& inputs,
                      std::vector<Output>* outputs) {
  return ops::BuildWhileLoop(scope.NewSubScope("inner"), inputs,
                             LessThanTenCond, AddOneBody, "inner_loop",
                             outputs);
}

TEST(ValidateControlFlowTest, InputsFromDifferentFrames) {
  Scope scope = Scope::NewRootScope().ExitOnError();
  std::vector<Output> inputs;
  inputs.push_back(ops::Placeholder(scope, DT_INT32));
  std::vector<Output> outputs;
  TF_ASSERT_OK(ops::BuildWhileLoop(scope.NewSubScope("outer"), inputs,
                                   LessThanTenCond, NestedLoopBody,
                                   "outer_loop", &outputs));
  std::unique_ptr<Graph> graph(new Graph(OpRegistry::Global()));
  TF_ASSERT_OK(scope.ToGraph(graph.get()));
  // {inner/Enter', 'outer/Switch'} --> 'inner/Merge'. 'inner/Enter' is in frame
  // 'inner_loop'. 'outer/Switch' is in frame 'outer_loop'.
  std::vector<ControlFlowInfo> info;
  Status status = BuildControlFlowInfo(graph.get(), &info);
  EXPECT_FALSE(status.ok());
  EXPECT_TRUE(absl::StrContains(status.error_message(),
                                "has inputs from different frames"))
      << status.error_message();
  EXPECT_TRUE(absl::StrContains(status.error_message(),
                                "{{node outer/body/inner/Merge}}"))
      << status.error_message();
  EXPECT_TRUE(absl::StrContains(status.error_message(),
                                "{{node outer/body/inner/Enter}}"))
      << status.error_message();
  EXPECT_TRUE(
      absl::StrContains(status.error_message(), "{{node outer/Switch}}"))
      << status.error_message();
}

TEST(ValidateControlFlowTest, MismatchedParentFrames) {
  Scope scope = Scope::NewRootScope().ExitOnError();
  std::vector<Output> inputs;
  inputs.push_back(ops::Placeholder(scope, DT_INT32));
  std::vector<Output> outputs;
  TF_ASSERT_OK(ops::BuildWhileLoop(scope, inputs, LessThanTenCond, AddOneBody,
                                   "test_loop", &outputs));
  std::unique_ptr<Graph> graph(new Graph(OpRegistry::Global()));
  TF_ASSERT_OK(scope.ToGraph(graph.get()));
  Node* enter_1 = nullptr;
  for (Node* node : graph->op_nodes()) {
    if (IsEnter(node)) {
      enter_1 = node;
    }
  }
  ASSERT_TRUE(enter_1 != nullptr);

  NodeDef enter;
  enter.set_name("Enter2");
  enter.set_op("Enter");
  (*enter.mutable_attr())["T"].set_type(DT_INT32);
  (*enter.mutable_attr())["frame_name"].set_s("test_loop");
  *enter.add_input() = "Enter";
  Status status;
  Node* enter_2 = graph->AddNode(enter, &status);
  TF_ASSERT_OK(status);
  graph->AddControlEdge(enter_1, enter_2);

  // SOURCE("") --> Enter("test_loop") --> Enter2("test_loop")
  // For node 'Enter', the parent frame of "test_loop" is empty.
  // For node 'Enter2', the parent frame of "test_loop" is "test_loop".
  std::vector<ControlFlowInfo> info;
  status = BuildControlFlowInfo(graph.get(), &info);
  EXPECT_FALSE(status.ok());
  EXPECT_TRUE(
      absl::StrContains(status.error_message(), "Mismatched parent frames"))
      << status.error_message();
  EXPECT_TRUE(absl::StrContains(status.error_message(), "{{node Enter2}}"))
      << status.error_message();
}

TEST(ValidateControlFlowTest, TwoLoopCond) {
  // Test that one frame has at most one LoopCond node. This is necessary for
  // functionalize control flow.
  Scope scope = Scope::NewRootScope().ExitOnError();
  std::vector<Output> inputs;
  inputs.push_back(ops::Placeholder(scope, DT_INT32));
  std::vector<Output> outputs;
  TF_ASSERT_OK(ops::BuildWhileLoop(scope, inputs, LessThanTenCond, AddOneBody,
                                   "test_loop", &outputs));
  outputs.clear();
  TF_ASSERT_OK(ops::BuildWhileLoop(scope.NewSubScope("sub"), inputs,
                                   LessThanTenCond, AddOneBody, "test_loop",
                                   &outputs, false));
  std::unique_ptr<Graph> graph(new Graph(OpRegistry::Global()));
  TF_ASSERT_OK(scope.ToGraph(graph.get()));
  std::vector<ControlFlowInfo> info;
  Status status = BuildControlFlowInfo(graph.get(), &info);
  EXPECT_FALSE(status.ok());
  EXPECT_TRUE(
      absl::StrContains(status.error_message(), "more than one LoopCond node"))
      << status.error_message();
  EXPECT_TRUE(
      absl::StrContains(status.error_message(), "{{node sub/LoopCond}}"))
      << status.error_message();
  EXPECT_TRUE(absl::StrContains(status.error_message(), "{{node LoopCond}}"))
      << status.error_message();
}

}  // namespace
}  // namespace tensorflow
